app/assets/javascripts/oxalis/view/action-bar/merge_modal_view.js

```diff
@@ -101,98 +108,91 @@ class MergeModalView extends PureComponent<Props, MergeModalViewState> {
-    }
-  };
-
   handleBeforeUploadNML = () => {
     this.setState({ isUploading: true });
   };
 
-  handleMergeTaskType = (event: SyntheticInputEvent<>) => {
-    event.preventDefault();
-    const { selectedTaskType } = this.state;
-    if (selectedTaskType != null) {
-      const url =
-        `/api/annotations/CompoundTaskType/${selectedTaskType}/` +
-        `merge/${this.props.tracingType}/${this.props.annotationId}`;
-      this.merge(url);
-    }
-  };
-
-  handleMergeProject = (event: SyntheticInputEvent<>) => {
+  handleMergeProject = async (event: SyntheticInputEvent<>, isLocalMerge: boolean) => {
     event.preventDefault();
     const { selectedProject } = this.state;
     if (selectedProject != null) {
-      const url =
-        `/api/annotations/CompoundProject/${selectedProject}/merge/` +
-        `${this.props.tracingType}/${this.props.annotationId}`;
-      this.merge(url);
+      if (isLocalMerge) {
+        const annotation = await getAnnotationInformation(selectedProject, "CompoundProject");
+        this.mergeAnnotationIntoActiveTracing(annotation);
+      } else {
+        const url =
+          `/api/annotations/CompoundProject/${selectedProject}/merge/` +
+          `${this.props.tracingType}/${this.props.annotationId}`;
+        this.merge(url);
+      }
     }
   };
 
-  handleMergeExplorativeAnnotation = async (event: SyntheticInputEvent<>) => {
+  handleMergeExplorativeAnnotation = async (
+    event: SyntheticInputEvent<>,
+    isLocalMerge: boolean,
+  ) => {
     event.preventDefault();
     const { selectedExplorativeAnnotation } = this.state;
 
     if (selectedExplorativeAnnotation != null) {
-      const url =
-        `/api/annotations/Explorational/${selectedExplorativeAnnotation}/merge/` +
-        `${this.props.tracingType}/${this.props.annotationId}`;
-      this.merge(url);
+      if (isLocalMerge) {
+        const annotation = await getAnnotationInformation(
+          selectedExplorativeAnnotation,
+          "Explorational",
+        );
+        this.mergeAnnotationIntoActiveTracing(annotation);
+      } else {
+        const url =
+          `/api/annotations/Explorational/${selectedExplorativeAnnotation}/merge/` +
+          `${this.props.tracingType}/${this.props.annotationId}`;
+        this.merge(url);
+      }
     }
   };
 
+  async mergeAnnotationIntoActiveTracing(annotation: APIAnnotationType): Promise<void> {
+    const tracing = await getTracingForAnnotation(annotation);
+    if (tracing.trees) {
+      const { trees, treeGroups } = tracing;
+      this.setState({ isUploading: true });
+      // Wait for an animation frame so that the loading animation is kicked off
+      await Utils.animationFrame();
+      this.props.addTreesAndGroupsAction(createTreeMapFromTreeArray(trees), treeGroups || []);
+      this.setState({ isUploading: false });
+      Toast.success(messages["tracing.merged"]);
+    } else {
+      Toast.error("Merging is not supported for volume tracings.");
```

Comment:
Should this be moved to the messages file as well? Actually, I'm not sure what our current practice regarding the messages.js is anyways? Maybe we should discuss this, not as part of this PR :)